### PR TITLE
Train Resnet Bounty [WIP] 

### PR DESCRIPTION
**Pull request by pythonlearner1025**
**Status** \
Log of one training step on 1xA100 GPU at BS=16, expected 62 hrs training time: 2.8x off. \
```text 56.88 ms run,  55.70 ms python,  64.61 ms fetch data, 1.18 ms prefetch data, 7.17 loss, 3.53 GB used,  1662.67 GFLOPS```\
Goal is to get GPU time <=20ms and pull data loading time below <=20ms. That should get us to below 24 hrs. I overlapped data loading and compute, cutting time in ~half from @ojaffe's attempt. Below is plan for cutting GPU & data times, will go down list for each until <24hrs. 

**GPU**
- BEAM search opt
- multi gpu 

**Data Loading**
- Preprocess on GPU using Tinygrad
- Preprocess on GPU using NVIDIA DALI (can I use DALI?)

### CODE REVIEW

**Comment by geohot** [comment]
Bounty locked, good luck!

**Comment by pythonlearner1025** [comment]
**UPDATE**
Depending on the CPU and I/O speed of the VM, <24hrs should be possible. With BEAM=4 on a A100, one step of resnet50 takes <50ms at BS=64. At BS=64 you need <80ms per train step to get <24hrs (at BS=128 160ms, etc), where each step time is given by max(dataload tm, gpu tm). So, if the dataload tm is below 80ms, you'll be under 24hrs.

First try running ```model_train.py``` with  ```benchmark_dataload_tm()``` which downloads & benchmarks dataload tm on the smaller imagenette2 dataset. If none of the batch sizes gets you <24hrs, don't waste time and $ doing the full run. You're limited by CPU & I/O.

Then do full run:
- ```export MODEL=resnet```
- ```export CUDA=1```
- ```export BEAM=4```
- ```export IMGNET_TRAIN=1```
- ```python extra/datasets/imagenet_download.py``` wait...
- ```python examples/mlperf/model_train.py``` wait for jit with beam...

If limited by dataloading, one solution would be to get a faster VM (fast CPU & SSD). VMs I rented on RunPod were super slow and didn't pass - for reference on my m1 mac running benchmark_dataload.py gives:
```
benchmarking dataload tm
**BS=16 W=4**
  79.96 hrs total tm   75.43hrs train tm    4.53hrs val tm
batch:  266.31 ms avg read   67.83 ms median read 2049.06 ms max read
unit:    9.30 ms avg read    6.48 ms median read   63.05 ms max read
**BS=16 W=8**
  81.93 hrs total tm   77.30hrs train tm    4.64hrs val tm
batch:  327.60 ms avg read   69.51 ms median read 3265.29 ms max read
unit:   13.93 ms avg read   10.38 ms median read  135.17 ms max read
**BS=16 W=16**
  84.61 hrs total tm   79.82hrs train tm    4.79hrs val tm
batch:  481.69 ms avg read   71.78 ms median read 7204.52 ms max read
unit:   25.58 ms avg read   15.20 ms median read  275.69 ms max read
**BS=32 W=4**
  44.77 hrs total tm   42.24hrs train tm    2.53hrs val tm
batch:  435.07 ms avg read   75.96 ms median read 7204.52 ms max read
unit:   18.59 ms avg read    7.89 ms median read  275.69 ms max read
**BS=32 W=8**
  46.82 hrs total tm   44.17hrs train tm    2.65hrs val tm
batch:  425.16 ms avg read   79.43 ms median read 7204.52 ms max read
unit:   16.95 ms avg read    9.72 ms median read  275.69 ms max read
**BS=32 W=16**
  49.25 hrs total tm   46.46hrs train tm    2.79hrs val tm
batch:  499.72 ms avg read   83.55 ms median read 7901.66 ms max read
unit:   23.81 ms avg read   12.18 ms median read  466.72 ms max read
**BS=64 W=4**
  25.31 hrs total tm   23.88hrs train tm    1.43hrs val tm
batch:  477.12 ms avg read   85.89 ms median read 7901.66 ms max read
unit:   18.56 ms avg read    8.05 ms median read  466.72 ms max read
**BS=64 W=8**
  25.92 hrs total tm   24.45hrs train tm    1.47hrs val tm
batch:  471.05 ms avg read   87.96 ms median read 7901.66 ms max read
unit:   17.06 ms avg read    9.50 ms median read  466.72 ms max read
**BS=64 W=16**
  27.77 hrs total tm   26.19hrs train tm    1.57hrs val tm
batch:  518.98 ms avg read   94.22 ms median read 7901.66 ms max read
unit:   21.54 ms avg read   11.07 ms median read  466.72 ms max read
**BS=128 W=4**
  16.07 hrs total tm   15.16hrs train tm    0.91hrs val tm
batch:  512.15 ms avg read  109.09 ms median read 7901.66 ms max read
unit:   17.46 ms avg read    8.51 ms median read  466.72 ms max read
**BS=128 W=8**
  18.28 hrs total tm   17.25hrs train tm    1.03hrs val tm
batch:  516.86 ms avg read  124.08 ms median read 7901.66 ms max read
unit:   16.40 ms avg read    9.83 ms median read  466.72 ms max read
**BS=128 W=16**
  19.35 hrs total tm   18.25hrs train tm    1.10hrs val tm
batch:  637.23 ms avg read  131.31 ms median read 16860.07 ms max read
unit:   19.59 ms avg read   11.20 ms median read  466.72 ms max read
RANK 0: Under 24hrs=True BS=128 W=4   15.16 hrs NOTE: if GPU tm under  162.89 ms, then under 24hrs
RANK 1: Under 24hrs=True BS=128 W=8   17.25 hrs NOTE: if GPU tm under  162.89 ms, then under 24hrs
RANK 2: Under 24hrs=True BS=128 W=16   18.25 hrs NOTE: if GPU tm under  162.89 ms, then under 24hrs
RANK 3: Under 24hrs=True BS=64 W=4   23.88 hrs NOTE: if GPU tm under   81.44 ms, then under 24hrs
RANK 4: Under 24hrs=False BS=64 W=8   24.45 hrs NOTE: no use in training, limited by cpu
RANK 5: Under 24hrs=False BS=64 W=16   26.19 hrs NOTE: no use in training, limited by cpu
RANK 6: Under 24hrs=False BS=32 W=4   42.24 hrs NOTE: no use in training, limited by cpu
RANK 7: Under 24hrs=False BS=32 W=8   44.17 hrs NOTE: no use in training, limited by cpu
RANK 8: Under 24hrs=False BS=32 W=16   46.46 hrs NOTE: no use in training, limited by cpu
RANK 9: Under 24hrs=False BS=16 W=4   75.43 hrs NOTE: no use in training, limited by cpu
RANK 10: Under 24hrs=False BS=16 W=8   77.30 hrs NOTE: no use in training, limited by cpu
RANK 11: Under 24hrs=False BS=16 W=16   79.82 hrs NOTE: no use in training, limited by cpu
```

**Comment by geohot** [comment]
You want access to a tinybox?

File: .gitignore [hunk 1]
```diff
@@ -5,6 +5,7 @@ notebooks
 .*.swp
 .*.swo
 *.pyc
+*.DS_Store
```
Comment (eichenroth) [inline .gitignore:8]: imo, this should be in your personal .gitignore file

**Pull request closed without merging**